From aaa9ed2dc16c67c576721e3d7951e4c265df866b Mon Sep 17 00:00:00 2001
From: Jason Wessel <jason.wessel@windriver.com>
Date: Thu, 16 Jul 2020 20:11:07 -0700
Subject: [PATCH] Revert "Revert "mmc: sdhci: Mask "spurious" interrupts""

This reverts commit a8913cf2ebc7a5bda92b2737271f8869741f6b18 in order
to fix the problem with small mmc cards where the kernel is printing
the message:

mmc0: Got data interrupt 0x00000002 even though no data operation was in progress.

Upstream-Status: Pending

Signed-off-by: Jason Wessel <jason.wessel@windriver.com>
---
 drivers/mmc/host/sdhci.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/drivers/mmc/host/sdhci.c b/drivers/mmc/host/sdhci.c
index 070f7dc42294..2e30fca8b44c 100644
--- a/drivers/mmc/host/sdhci.c
+++ b/drivers/mmc/host/sdhci.c
@@ -3104,6 +3104,10 @@ static irqreturn_t sdhci_irq(int irq, void *dev_id)
 			result = IRQ_WAKE_THREAD;
 		}
 
+		if ((intmask & SDHCI_INT_DATA_END) && !host->data &&
+		    host->cmd && (host->cmd == host->cmd->mrq->stop))
+			intmask &= ~SDHCI_INT_DATA_END;
+
 		if (intmask & SDHCI_INT_CMD_MASK)
 			sdhci_cmd_irq(host, intmask & SDHCI_INT_CMD_MASK, &intmask);
 
-- 
2.17.1

